libsolutil/CDCL.cpp

```diff
@@ -230,9 +245,40 @@ std::pair<Clause, size_t> CDCL::analyze(Clause _conflictClause)
 	return {move(learntClause), backtrackLevel};
 }
 
-void CDCL::addClause(Clause _clause)
+void CDCL::addClause(const Clause& _lits)
 {
-	m_clauses.push_back(make_unique<Clause>(move(_clause)));
+	if (!ok) return;
+
+	Clause clause{_lits};
+	Clause clause_updated;
+	for (const auto& l: clause)
+	{
+		// Clause is satisfied, nothing to do.
+		if (value(l) == TriState::t_true())
+			return;
+
+		// Remove literal from clause.
```

Comment:
Why can I do that? Is this only for learnt clauses? If yes, we should rename the function

Reply:
We can do this because if there is a literal that's already satisfied according to the current assignment, then the clause is already satisfied, so it has no effect.

This can occur if we have two clauses that come in, in sequence:

```
1 0
1 2 0
```

In this case, the 2nd clause can be removed, because variable `1` is already set to `TriState{true}` so the clause has no effect, it's already satisfied by literal `1`. Conversely, in the case of:

```
1 0
-1 2 3 0
```

The 2nd clause will be simplified to `2 3 0`, since `1` is already `TriState{true}`, and so the literal `-1` evaluates to `TriState{false}` and has no effect.

Reply:
But again, is this only for learnt clauses?

Reply:
Ah sorry. With the new system, `addClause` is not called for learnt clauses at all. We deal with 1- and 2- long learnt clauses post-analysis, and the larger ones we attach. So the invariant holds for all clauses, but this code is not called for learnt clauses.